File: day19/day19.go
package day19

import (
	"fmt"
	"strings"
	"sync"
)

var (
	cache   = make(map[string]int)
	cacheMu sync.Mutex
)

func Day19(s string) {
	patterns, partials := parse(s)
	res := 0
	perms := 0
	resChan := make(chan int, len(patterns))
	var wg sync.WaitGroup
	var wwg sync.WaitGroup

	wwg.Add(1)
	go func() {
		defer wwg.Done()
		for val := range resChan {
			if val > 0 {
				res++
			}
			perms += val
		}
	}()

	for _, line := range patterns {
		wg.Add(1)
		go func(line string) {
			defer wg.Done()
			resChan <- dp(partials, line)
		}(line)
	}

	wg.Wait()
	close(resChan)
	wwg.Wait()

	fmt.Println("Part 1")
	fmt.Println(res)
	fmt.Println("Part 2")
	fmt.Println(perms)
}

func dp(partials []string, line string) int {
	cacheMu.Lock()
	if result, ok := cache[line]; ok {
		cacheMu.Unlock()
		return result
	}
	cacheMu.Unlock()

	if len(line) == 0 {
		return 1
	}

	res := 0
	for _, k := range partials {
		if strings.HasPrefix(line, k) {
			res += dp(partials, line[len(k):])
		}
	}

	cacheMu.Lock()
	cache[line] = res
	cacheMu.Unlock()

	return res
}

func parse(s string) ([]string, []string) {
	split := strings.Split(s, "\n\n")
	lines := strings.Split(strings.TrimSpace(split[1]), "\n")
	return lines, strings.Split(split[0], ", ")
}

File: day19/day19.rs
use std::collections::HashMap;
use std::fs;
use std::io::{self, Read};
use std::time::Instant;

fn main() {
    let start = Instant::now();
    let filename = "input.txt";
    match read_input_from_file(filename) {
        Ok(input) => {
            let (partials, patterns) = parse(&input);
            let mut cache: HashMap<String, i64> = HashMap::new();

            let mut res = 0;
            let mut perms = 0;

            for line in patterns {
                let temp = dp(&partials, &line, &mut cache);

                if temp > 0 {
                    res += 1;
                }

                perms += temp;
            }

            println!("Part 1\n{}", res);
            println!("Part 2 \n{}", perms);
            println!("{:?}", start.elapsed())
        }
        Err(e) => eprintln!("Error reading file: {}", e),
    }
}

fn dp(partials: &[String], line: &str, cache: &mut HashMap<String, i64>) -> i64 {
    if let Some(&result) = cache.get(line) {
        return result;
    }

    if line.is_empty() {
        return 1;
    }

    let mut res = 0;

    for k in partials {
        if line.starts_with(k) {
            res += dp(partials, &line[k.len()..], cache);
        }
    }

    cache.insert(line.to_string(), res);
    res
}

fn parse(input: &str) -> (Vec<String>, Vec<String>) {
    let parts: Vec<&str> = input.split("\n\n").collect();
    let partials: Vec<String> = parts[0].split(", ").map(|s| s.to_string()).collect();
    let patterns: Vec<String> = parts[1]
        .lines()
        .map(|line| line.trim().to_string())
        .collect();

    (partials, patterns)
}

fn read_input_from_file(filename: &str) -> io::Result<String> {
    let mut file = fs::File::open(filename)?;
    let mut content = String::new();
    file.read_to_string(&mut content)?;
    Ok(content)
}

